noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays/assert_split_sorted_transformed_value_arrays.nr

```diff
@@ -0,0 +1,543 @@
+use crate::{
+    abis::side_effect::Ordered, traits::{Empty, is_empty},
+    utils::arrays::{array_length, sort_get_split_order_hints::SplitOrderHints, validate_array}
+};
+
+// original_array must be valid, i.e. validate_array(original_array) == true
+// transformed_value_array must be verified against original_array before calling this function.
+fn assert_split_sorted_transformed_value_arrays<T, S, N>(
+    original_array: [T; N],
+    transformed_value_array: [S; N],
+    split_counter: u32,
+    sorted_transformed_value_array_lt: [S; N], // Values whose counters are less than the split counter.
+    sorted_transformed_value_array_gte: [S; N], // Values whose counters are greater than or equal to the split counter.
+    sorted_counters_lt: [u32; N], // Counters of the values in sorted_transformed_value_array_lt.
+    sorted_counters_gte: [u32; N], // Counters of the values in sorted_transformed_value_array_gte.
+    index_hints: [u32; N], // The index of the item in the correspinding sorted_transformed_value_array_(lt/gte) for each item in the original_array.
+    ascending: bool // Whether the items in sorted_transformed_value_array_(lt/gte) is in ascending order.
+) where T: Ordered + Empty + Eq, S: Empty + Eq {
+    // Can use array_length instead of validate_array for the original_array because it's taken from the previous kernel and guaranteed to be valid.
+    let total_num = array_length(original_array);
+
+    let mut num_lt = 0;
+    let mut num_gte = 0;
+    let mut should_check = true;
+    for i in 0..N {
+        should_check &= i != total_num;
+        if should_check {
+            let original = original_array[i];
+            let value = transformed_value_array[i];
+            let sorted_index = index_hints[i];
+            let is_lt = (original.counter() < split_counter) | (original.counter() == 0); // If counter is 0, the value should always be in the lt(non-revertible) set. Currently this only applies to the first nullifier.
```

Comment:
I thought about this a bit and I think it would fail the below checks in the `if num != 0` block - we would have two things with counter = 0 (the first nullifier, which I think is always added pre this call, and the 'bad' thing added by the app contract). So once we deal with one of those things, `num` gets incremented, and it will fail the `sorted_counters` check.
I don't think you could get around it due to the inc/dec checks on `sorted_counters` but I could be missing something!

Reply:
@sirasistant as discussed - my comment above only applies to the nullifiers array, different arrays would not contain the first nullifier so they are still vulnerable to this. Good catch!